### ISSUE TITLE
A problem with Geneagrapher backend

### ISSUE THREAD

**Issue by johnshadid** [davidalber/geneagrapher-notebook-feedback#1]
There appears to be a problem with the Geneagrapher backend.
This is apparent in 2 ways.

1) using the website version of the tool on the page https://github.com/davidalber/geneagrapher
I have tried to run the example for Carl Guass with  ID 18231. The tool does not complete and produce a graph.
This has been the case since the weekend.

2) I have just downloaded the ggrapher tool and run this example and I get this error.

/usr/local/anaconda3/bin/ggrapher -o guass_2023.dot 18231:a
Geneagrapher backend is currently unavailable.

If this problem persists, please create an issue at
https://github.com/davidalber/geneagrapher/issues/new, and include the
following in the issue body:

    Message: Geneagrapher backend is currently unavailable.
    Command: /usr/local/anaconda3/bin/ggrapher -o guass_2023.dot 18231:a

Thank you for creating this tool and making it available!

**Comment by davidalber** [comment]
Thank you for the detailed report, @johnshadid!

**tl;dr** The service is running again.

# Diagnosis
The service log was capturing exceptions like this:
```
2023-11-28 17:34:32.520 [geneagrapher_live] web.1 {"dt":"2023-11-28 17:34:32.520126","kind":"cache hit","record_id":125154,"session_id":"5a23e903-c757-4371-a913-8f0de483d64b"}
2023-11-28 17:34:32.521 [geneagrapher_live] web.1 2023-11-28 17:34:32.520891 5a23e903-c757-4371-a913-8f0de483d64b connection handler failed
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   + Exception Group Traceback (most recent call last):
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |   File "/app/.heroku/python/lib/python3.11/site-packages/websockets/legacy/server.py", line 236, in handler
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |     await self.ws_handler(self)
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |   File "/app/geneagrapher_live/app.py", line 287, in handle
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |     await do_build_graph(ws, request)
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |   File "/app/geneagrapher_live/app.py", line 265, in do_build_graph
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |     return await do_build_graph_work(ws, start_nodes, request.get("options", {}))
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |   File "/app/geneagrapher_live/app.py", line 216, in do_build_graph_work
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |     graph = await build_graph(
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |             ^^^^^^^^^^^^^^^^^^
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |   File "/app/.heroku/python/lib/python3.11/site-packages/geneagrapher_core/traverse.py", line 235, in build_graph
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |     async with asyncio.TaskGroup() as tg:
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |   File "/app/.heroku/python/lib/python3.11/asyncio/taskgroups.py", line 147, in __aexit__
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   |     raise me from None
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
2023-11-28 17:34:32.521 [geneagrapher_live] web.1   +-+---------------- 1 ----------------
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     | Traceback (most recent call last):
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     |   File "/app/.heroku/python/lib/python3.11/site-packages/geneagrapher_core/traverse.py", line 221, in fetch_and_process
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     |     await add_neighbor_work(record, td)
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     |   File "/app/.heroku/python/lib/python3.11/site-packages/geneagrapher_core/traverse.py", line 199, in add_neighbor_work
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     |     for id in record[key]:
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     |               ~~~~~~^^^^^
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     | KeyError: 'advisors'
2023-11-28 17:34:32.521 [geneagrapher_live] web.1     +------------------------------------
2023-11-28 17:34:32.522 [geneagrapher_live] web.1 2023-11-28 17:34:32.522033 5a23e903-c757-4371-a913-8f0de483d64b connection closed
```

That suggested something was wrong with the cached record for Record 125154. Looking in Redis confirmed that.

```
> get ggrapher:record:1:125154
{}
```

That would cause the exception in the log.

# Fix
I deleted that record from the cache, and then the service got hung up on another cached record. Then another, and then another. I ended up flushing the entire cache, and now Geneagrapher is working again.

## CLI Confirmation
Before:
```
$ ggrapher 18231:a
Geneagrapher backend is currently unavailable.██████████:::::::::::::::] 12/16
```

After:
```
$ ggrapher 18231:a
digraph {
    graph [ordering="out"];
    node [shape=plaintext];
    edge [style=bold];

    18230 [label="Johann Friedrich Pfaff\nGeorg-August-Universität Göttingen (1786)"];
.
.
.
```

## Observable Notebook Confirmation
Before:
![image](https://github.com/davidalber/geneagrapher-notebook-feedback/assets/933552/a30f5dbb-f648-4ec0-9667-2e413c93b1a6)

☝️ It just hung there like that.

After:
<img width="1180" alt="image" src="https://github.com/davidalber/geneagrapher-notebook-feedback/assets/933552/a08d1f56-8033-4e09-87c6-8ee8fa7b3f43">

# The Issue
It's not apparent how (at least) several corrupt values were cached. However, nightly tests that use the live Math Genealogy Project site failed intermittently earlier this week. I took their failure to mean that the Math Genealogy Project site was having issues. It's possible that Geneagrapher retrieved data from the site during a partial outage and cached bad information. Here is what [one of those failures](https://github.com/davidalber/geneagrapher-core/actions/runs/6994666930/job/19028688008) looked like:

```
=================================== FAILURES ===================================
___________________________ test_still_works[18231] ____________________________

record_id = 18231

    @pytest.mark.asyncio
    @pytest.mark.live
    @pytest.mark.parametrize("record_id", (18231,))
    async def test_still_works(record_id: int) -> None:
        """Compare the computed record for a given `record_id` to stored data."""
        _, expected = load_record_test(str(record_id))
    
        # Clean the test data up a bit to make it look like a record.
        del expected["is_valid"]
        expected["id"] = record_id
    
>       assert await get_record(RecordId(record_id)) == expected
E       AssertionError: assert None == {'advisors': [18230], 'descendants': [18603, 18233, 62547, 29642, 55175, 225908, ...], 'id': 18231, 'institution': 'Universität Helmstedt', ...}

tests/test_live.py:22: AssertionError
```

It will take more investigation to reproduce a likely cause for what happened here.
